harvardkey_cas/backends.py

```diff
@@ -30,30 +30,33 @@ def authenticate(self, request, ticket, service):
         """
 
         client = get_cas_client(service_url=service)
+        logger.debug(f'CAS client type: {type(client)}')
         username, attributes, pgtiou = client.verify_ticket(ticket)
-        if attributes and request:
-            request.session['user_attributes'] = attributes
-            logger.debug('fetched user attributes from CAS %s' % (attributes))
-
-        authenticationType = attributes.get('authenticationType')
-        logger.debug('authenticationType =%s' % (authenticationType))
 
         if not username:
+            logger.warn(f"no username returned by CAS server for ticket {ticket}")
             return None
 
+        if attributes and request:
+            request.session['user_attributes'] = attributes
+            logger.debug(f'fetched user attributes from CAS: {attributes}')
+
+            authentication_type = attributes.get('authenticationType')
+            logger.debug(f'authenticationType = {authentication_type}')
+        else:
+            logger.warn(f'no attributes found in CAS response for username {username} and ticket {ticket}')
```

Comment:
This could technically happen if `request` was falsy too -- do we know why we were checking `request`, other than to ensure the `request.session` lookup wasn't going to throw a 500 error? (Seems like a strange thing to validate in the first place -- in practice, I'd imagine we could remove the `request` check on line 40, or else put the `request.session` in its own try/catch or conditional block.)